src/sparse_blas/backends/rocsparse/rocsparse_handles.hpp

```diff
@@ -0,0 +1,105 @@
+                                                                         size) {}
+};
+
+struct dense_matrix_handle : public detail::generic_dense_matrix_handle<rocsparse_dnmat_descr> {
+    template <typename T>
+    dense_matrix_handle(rocsparse_dnmat_descr roc_descr, T* value_ptr, std::int64_t num_rows,
+                        std::int64_t num_cols, std::int64_t ld, layout dense_layout)
+            : detail::generic_dense_matrix_handle<rocsparse_dnmat_descr>(
+                  roc_descr, value_ptr, num_rows, num_cols, ld, dense_layout) {}
+
+    template <typename T>
+    dense_matrix_handle(rocsparse_dnmat_descr roc_descr, const sycl::buffer<T, 1> value_buffer,
+                        std::int64_t num_rows, std::int64_t num_cols, std::int64_t ld,
+                        layout dense_layout)
+            : detail::generic_dense_matrix_handle<rocsparse_dnmat_descr>(
+                  roc_descr, value_buffer, num_rows, num_cols, ld, dense_layout) {}
+};
+
+struct matrix_handle : public detail::generic_sparse_handle<rocsparse_spmat_descr> {
+    // A matrix handle should only be used once per operation to be safe with the rocSPARSE backend.
+    // An operation can store information in the handle. See details in https://github.com/ROCm/rocSPARSE/issues/332.
+private:
+    bool used = false;
+
+public:
+    template <typename fpType, typename intType>
+    matrix_handle(rocsparse_spmat_descr roc_descr, intType* row_ptr, intType* col_ptr,
+                  fpType* value_ptr, detail::sparse_format format, std::int64_t num_rows,
+                  std::int64_t num_cols, std::int64_t nnz, oneapi::mkl::index_base index)
+            : detail::generic_sparse_handle<rocsparse_spmat_descr>(
+                  roc_descr, row_ptr, col_ptr, value_ptr, format, num_rows, num_cols, nnz, index) {}
+
+    template <typename fpType, typename intType>
+    matrix_handle(rocsparse_spmat_descr roc_descr, const sycl::buffer<intType, 1> row_buffer,
+                  const sycl::buffer<intType, 1> col_buffer,
+                  const sycl::buffer<fpType, 1> value_buffer, detail::sparse_format format,
+                  std::int64_t num_rows, std::int64_t num_cols, std::int64_t nnz,
+                  oneapi::mkl::index_base index)
+            : detail::generic_sparse_handle<rocsparse_spmat_descr>(roc_descr, row_buffer,
+                                                                   col_buffer, value_buffer, format,
+                                                                   num_rows, num_cols, nnz, index) {
+    }
+
+    void check_valid_handle(const std::string& function_name) {
+        if (used) {
+            throw mkl::unimplemented(
+                "sparse_blas", function_name,
+                "The backend does not support re-using the same sparse matrix handle in multiple operations.");
+        }
+        if (this->format == detail::sparse_format::COO &&
+            !this->has_matrix_property(matrix_property::sorted)) {
```

Comment:
In the documentation, you've written:
> - The CSR format requires the column indices to be sorted within each row.

So shouldn't you have `if ((this->format == detail::sparse_format::COO || this->format == detail::sparse_format::CSR) && !this->has_matrix_property(matrix_property::sorted))` here?

Reply:
Good point, I added this check in https://github.com/oneapi-src/oneMKL/pull/544/commits/b6518c90d986b7d65f37aed368fcaeac18adad79
I've also made sure we also have tests on AMD and Nvidia which are skipped when a property is not set. The runtime example also need to set this property now.